src/node/txreconciliation.cpp

```diff
@@ -363,79 +363,75 @@ class TxReconciliationTracker::Impl
+        // transaction we are picking the peers for.
+        CSipHasher deterministic_randomizer_in{m_deterministic_randomizer};
+        deterministic_randomizer_in.Write(wtxid.ToUint256());
+        CSipHasher deterministic_randomizer_out{deterministic_randomizer_in};
+        const size_t inbounds_target_size = ((deterministic_randomizer_in.Finalize() & 0xFFFFFFFF) + uint64_t(n * 0x100000000)) >> 32;
+
+        // Pick all reconciliation registered peers and assign them a deterministically random value based on their peer id
+        // Also, split peers in inbounds/outbounds
+        std::vector<std::pair<uint64_t, NodeId>> sorted_inbounds, sorted_outbounds;
         for (const auto& [node_id, op_peer_state]: m_states) {
             const auto peer_state = std::get_if<TxReconciliationState>(&op_peer_state);
-            if (peer_state && peer_state->m_we_initiate == we_initiate) {
-                uint64_t hash_key = CSipHasher(deterministic_randomizer).Write(node_id).Finalize();
-                best_peers.emplace_back(hash_key, node_id);
+            if (peer_state) {
+                if (peer_state->m_we_initiate) {
+                    uint64_t hash_key = CSipHasher(deterministic_randomizer_out).Write(node_id).Finalize();
+                    sorted_outbounds.emplace_back(hash_key, node_id);
+                } else {
+                    uint64_t hash_key = CSipHasher(deterministic_randomizer_in).Write(node_id).Finalize();
+                    sorted_inbounds.emplace_back(hash_key, node_id);
+                }
+
             }
         }
 
-        std::sort(best_peers.begin(), best_peers.end());
-
-        auto it = best_peers.begin();
-        for (size_t i = 0; i < targets_size && it != best_peers.end(); ++i, ++it) {
-            if (it->second == peer_id) return true;
+        // Sort the peers based on their assigned random value, extract the node_ids and trim the collections to size
+        std::vector<NodeId> in_fanout_targets;
+        if (inbounds_target_size >= 1) {
+            std::sort(sorted_inbounds.begin(), sorted_inbounds.end());
+            for_each(sorted_inbounds.begin(), sorted_inbounds.end(),
+                    [&in_fanout_targets](auto& keyed_peer) { in_fanout_targets.push_back(keyed_peer.second); });
+            in_fanout_targets.resize(inbounds_target_size);
         }
-        return false;
+        std::vector<NodeId> out_fanout_targets;
+        if (outbounds_target_size >= 1) {
+            std::sort(sorted_outbounds.begin(), sorted_outbounds.end());
+            for_each(sorted_outbounds.begin(), sorted_outbounds.end(),
+                    [&out_fanout_targets](auto& keyed_peer) { out_fanout_targets.push_back(keyed_peer.second); });
+            out_fanout_targets.resize(outbounds_target_size);
+        }
+
+        return std::make_tuple(in_fanout_targets, out_fanout_targets);
```

Comment:
4234f5ebe133b949080aaa56da8cbdd18b650ff2

We can just merge the two vectors, no? There is no use for this separation.

Reply:
Yeah, you're right. Given how small the two vectors are, calling `ShouldFanoutTo` with a merged vector should not incur any major overhead, so merging will make it easier to read